### DISCUSSION TITLE
K3s read/writes and wear on flash storage such as NAND

### DISCUSSION THREAD

**Discussion by bettio** [k3s-io/k3s#7694] [General]
**Environmental Info:**
K3s Version:
x86_64 node:
```
# k3s -v
k3s version v1.22.5+k3s1 (405bf79d)
go version go1.16.10
```

Node(s) CPU architecture, OS, and Version:
```
x86_64 node: 
Linux myhost 5.4.0-96-generic #109-Ubuntu SMP Wed Jan 12 16:49:16 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

Cluster Configuration:
Just one node.

**Describe the bug:**
According to some tests that we did both on a k3s cluster running on an aarch64 node that runs a number of pods, and and a freshly installed instance on a x86_64 node, k3s performs an unexpected amount of writes to the disk.

On the freshly installed x86_64 node:
```
# pidstat -p 1811 -dl 60
Linux 5.4.0-96-generic (myhost)     01/28/22        _x86_64_        (4 CPU)

16:27:29      UID       PID   kB_rd/s   kB_wr/s kB_ccwr/s iodelay  Command
16:28:29        0      1811      0.00      6.33      0.00       0  /usr/local/bin/k3s server 
d16:29:29        0      1811      0.00      6.87      0.00       0  /usr/local/bin/k3s server 
16:30:29        0      1811      0.00      7.07      0.00       0  /usr/local/bin/k3s server 
16:31:29        0      1811      0.00     10.73      0.00       0  /usr/local/bin/k3s server 
16:32:29        0      1811      0.00      9.07      0.00       0  /usr/local/bin/k3s server 
e16:33:29        0      1811      0.00      8.13      0.00       0  /usr/local/bin/k3s server 
^C
Average:        0      1811      0.00      8.03      0.00       0  /usr/local/bin/k3s server 
```

This sums up to ~720 MiB/day, just for an "empty" freshly installed single node cluster.

On the k3s aarch64 master node, that is a cluster where we run 11 pods, we got:
```
pidstat -dl 60 -p 37662
Linux 5.15.13-1-aarch64-ARCH (otherhost)    01/28/2022      _aarch64_       (6 CPU)

03:15:20 PM   UID       PID   kB_rd/s   kB_wr/s kB_ccwr/s iodelay  Command
03:16:20 PM     0     37662      0.00     49.60      0.00       0  /usr/bin/k3s server
03:17:20 PM     0     37662      0.00     35.40      0.00       0  /usr/bin/k3s server                                                                                                        
03:18:20 PM     0     37662      0.00     39.53      0.00       0  /usr/bin/k3s server                                                                                                        
03:19:20 PM     0     37662      0.00     34.60      0.00       0  /usr/bin/k3s server                                                                                                        
03:20:20 PM     0     37662      0.00     44.53      0.00       0  /usr/bin/k3s server                                                                                                        
03:21:20 PM     0     37662      0.00     54.33      0.00       0  /usr/bin/k3s server                                                                                                        
03:22:20 PM     0     37662      0.00     36.07      0.00       0  /usr/bin/k3s server                                                                                                        
03:23:20 PM     0     37662      0.00     44.53      0.00       0  /usr/bin/k3s server 
```

That sums up to ~3.3 GiB/day.

I'm worried this would wear pretty quickly NAND, making k3s unsuitable for embedded systems that are meant to be up&running 24/7. 

**Steps To Reproduce:**

This already happens on a freshly installed node, such as the following x86_64 node:

```
# kubectl get nodes
NAME                 STATUS   ROLES                  AGE    VERSION
myhost   Ready    control-plane,master   101m   v1.22.5+k3s1
```

```
# kubectl get pods -A
NAMESPACE     NAME                                     READY   STATUS      RESTARTS   AGE
kube-system   local-path-provisioner-64ffb68fd-cnnsq   1/1     Running     0          92m
kube-system   coredns-85cb69466-vv2m5                  1/1     Running     0          92m
kube-system   metrics-server-9cf544f65-jcmq4           1/1     Running     0          92m
kube-system   helm-install-traefik-crd--1-sjgpb        0/1     Completed   0          92m
kube-system   helm-install-traefik--1-jl5vh            0/1     Completed   1          92m
kube-system   svclb-traefik-ntbqb                      2/2     Running     0          92m
kube-system   traefik-786ff64748-nqvx9
```

**Expected behavior:**

k3s should avoid excessive disk writes in order to be suitable on a truly embedded system, that might be installed on the field therefore replacing it might be costly. 

**Actual behavior:**

k3s keeps writing constantly to disk, wearing NAND storage.

**Additional context / logs:**

```
# lsof -p 1811 | grep REG  
k3s-serve 1811 root  txt       REG              252,1 131471184 258324 /var/lib/rancher/k3s/data/2e877cf4762c3c7df37cc556de3e08890fbf450914bb3ec042ad4f36b5a2413a/bin/containerd
k3s-serve 1811 root  mem-r     REG              252,1     32768 258578 /var/lib/rancher/k3s/server/db/state.db-shm
k3s-serve 1811 root    8ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root    9u      REG              252,1   5846312 258577 /var/lib/rancher/k3s/server/db/state.db-wal
k3s-serve 1811 root   10ur     REG              252,1     32768 258578 /var/lib/rancher/k3s/server/db/state.db-shm
k3s-serve 1811 root   11ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   15ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   38ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   39ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   43ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   55ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   59ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   61ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   62ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   63ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root   68ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root  147w      REG              252,1     40556 258661 /var/lib/rancher/k3s/agent/containerd/containerd.log
k3s-serve 1811 root  172ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root  173ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root  178ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root  180ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root  181ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root  182ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root  183ur     REG              252,1   2441216 258576 /var/lib/rancher/k3s/server/db/state.db
k3s-serve 1811 root  360u      REG              252,1   5846312 258577 /var/lib/rancher/k3s/server/db/state.db-wal
k3s-serve 1811 root  364u      REG              252,1   5846312 258577 /var/lib/rancher/k3s/server/db/state.db-wal
```
```
# file /var/lib/rancher/k3s/server/db/*
/var/lib/rancher/k3s/server/db/state.db:     SQLite 3.x database, last written using SQLite version 3034001
/var/lib/rancher/k3s/server/db/state.db-shm: data
/var/lib/rancher/k3s/server/db/state.db-wal: SQLite Write-Ahead Log, version 3007000
```
```
# du -k /var/lib/rancher/k3s/server/db/*
3204    /var/lib/rancher/k3s/server/db/state.db
32      /var/lib/rancher/k3s/server/db/state.db-shm
5712    /var/lib/rancher/k3s/server/db/state.db-wal
```

**Backporting**

- [ ] Needs backporting to older releases

**Comment by brandond** [comment]
This isn't really K3s specific; Kubernetes in general is quite write-heavy with regards to the datastore. Sqlite reduces the write overehad significantly compared to etcd (which also forces a full fsync on every write); I'm not sure how it could be reduced any further while still guaranteeing datastore consistency.

**Comment by j-landru** [comment]
Pending a possible consideration of the issue, on this type of support it could be recommended to deport the storage space /var/lib/rancher/k3s/server on a f2fs filesystem ?

**Comment by brandond** [comment]
Honestly I'm not sure that other bits of K3s (specifically, the containerd overlayfs snapshotter) would even work on f2fs. I would ensure that all the important bits are on a more traditional ext4/xfs/btrfs/etc - on a SSD or other low-latency storage medium, if possible.

**Comment by j-landru** [comment]
FYI, just to share my experience, in my attempt to build an IPv6 only cluster, on a K3OS VM, I overlayed a DELTA labeled f2fs block device to segregate persistent storage from a read-only rootfs. No problem on f2fs side, the main issue was containerd which doesn't seem to like to snaphot on a tow level nested overlayfs space.

- More info on : https://github.com/rancher/k3os/discussions/832
- Detailed howto on : https://framagit.org/-/snippets/5742

**Comment by shikoku1** [comment]
> This isn't really K3s specific

That's probably true, but, from the k3s website:
> K3s is a highly available, certified Kubernetes distribution designed for production workloads in **unattended, resource-constrained, remote locations** or inside **IoT appliances**.

We can agree that it's not k3s's direct fault, but we can probably also agree that writing several GBytes of data every day is not ideal for this application, and it's not what people expect when they read that an app is _designed for production workloads_ in "resource-constrained, remote locations". 

By the way, I've found a workaround using [this project](https://github.com/graysky2/anything-sync-daemon)

**Comment by brandond** [comment]
There are definitely varying definitions of "resource constrained". I don't think we're currently targeting, nor do we test on, systems that use NAND flash storage or flash-specific filesystems. The lower we currently target or test on is Raspberry Pi 3 class hardware.

**Comment by shikoku1** [comment]
but raspberry pi 3 uses a micro sd that has this exact problem, like most other sbcs

**Comment by brandond** [comment]
Yes, but I'm not aware of any supported distros that ship with f2fs by default.

MicroSD cards seem to work OK if you use one that has sufficient size to wear-level properly, and IO capacity that meeds the combined requirements of the datastore and workload. There are a couple issues on the tracker where users report success with various brands of SD cards; not all work well. For reliable operation, I would probably still recommend using a Pi4 or equivalent and booting off an external SSD or NVME disk.

**Comment by j-landru** [comment]
> Yes, but I'm not aware of any supported distros that ship with f2fs by default.

Yes in most distos, f2fs is not enabled by default and must be added as an extra. In my case, k3os distro, rootfs (K3OS_STATE block device) is ext4, only auxilliary (K3OS_DELTA) block device for persistent data (config, log, state...) is f2fs. My main goal is having the same unique immutable rootfs image shared by all cluster nodes. I don't want to update each node  indivudally. Update consists of building a new single reference image file and sequentially restart each node (maybe with kured) from the new reference image.

>  I would probably still recommend using a Pi4 or equivalent and booting off an external SSD or NVME disk.

My future plan is to deploy k3s on a picocluster (a set of 6 Marvell MacchiatoBIN sbc with sata SSD) and I would like to dedicate the entire SSD or nvme sbc storage for kube distributed storage (Rook/ceph or Rancher/longhorn or maybe open-io). that's why I plan to use auxilliary k3OS_DELTA bloc device on the sbc micro-sd, but the it may also be remote (iscsi or other remote bloc storage).

**Comment by mihaigalos** [comment]
What are the paths `k3s` writes to and what are their sizes? - I assume the quoted several GB/day are rotated at some point.

I'm considering writing state (i.e. `etcd`) either to an `nfs` mount (SSD-backed) or potentially even to `tmpfs`, provided it fits and the cluster contains at least a couple of nodes.

**Comment by j-landru** [comment]
At first sight, at least  `/var/lib/rancher` and `/var/log` and probably `/etc` if the persistence of node configurations is not ensured by another declarative mechanism (such as cloud init or terraform,...). What about `/home` or `/root` for ssh pub keys ?

**Comment by mihaigalos** [comment]
`/var/log` can already be a `tmpfs` by using [log2ram](https://github.com/azlux/log2ram).

> What about `/home` or `/root` for ssh pub keys ?

Ssh keys are static and don't change that often. I'm not concerned about wear of the i.e. microSD on a Pi there.

**Comment by mihaigalos** [comment]
I've written this Justfile (runnable with `just add_master` and `just add_worker`) to deploy k3s in docker.

Docker daemon logging is disabled (with `--log-driver=none`) and stdout and stderr redirected to `/var/log/k3s` which is a `tmpfs` ramdisk using log2ram.

I'm aware of `k3d` but that needs to run on the same docker engine. This implementation should run on multiple nodes, though that is currently untested.

It runs fine on 1 node (hosting master and worker):

```justfile
set shell := ["bash", "-uc"]

default:
    @just --list --unsorted

network := "k3s-home"

@_setup:
    [ -z $(sudo docker network ls --filter name=^{{ network }}$ --format="{{{{ .Name }}" ) ] && sudo docker network create k3s-home || true

add_master: _setup
    #!/bin/bash
    MASTER_NAME=k3s-master
    sudo docker run --rm -t \
        --hostname $MASTER_NAME \
        --log-driver=none \
        --name=$MASTER_NAME \
        --net={{ network }} \
        --privileged `# TODO: remove` \
        -v /var/log/k3s:/var/log \
        rancher/k3s:latest server > /var/log/k3s/$MASTER_NAME.log 2>&1 &
    disown

add_worker: _setup
    #!/bin/bash
    K3S_TOKEN=$(sudo docker exec -it k3s-master cat /var/lib/rancher/k3s/server/node-token | tr -d '\n' | tr -d '\r')
    WORKER_INDEX=$(( $(sudo docker ps | grep k3s-worker | wc -l) +1 ))
    WORKER_NAME=k3s-worker-$WORKER_INDEX
    sudo docker run --rm -t \
        --log-driver=none \
        --name=$WORKER_NAME \
        --net={{ network }} \
        --privileged `# TODO: remove` \
        --tmpfs /run \
        --tmpfs /var/run \
        -e K3S_NODE_NAME=$WORKER_NAME \
        -e K3S_TOKEN=$K3S_TOKEN \
        -e K3S_URL=https://k3s-master:6443 \
        -v /var/log/k3s:/var/log \
        rancher/k3s:latest agent > /var/log/k3s/$WORKER_NAME.log 2>&1 &
    disown
    sleep 3 # wait for registration of node in order to be able to wait for conditions pertaining to it.
    sudo docker exec -it k3s-master kubectl wait --for=condition=ready --timeout=300s node --all
    sudo docker exec -it k3s-master kubectl label node $WORKER_NAME kubernetes.io/role=worker

get_nodes:
    sudo docker exec -it k3s-master kubectl get nodes
```

**Comment by mihaigalos** [comment]
Seems however after a couple of minutes the worker becomes `NotReady`:
```bash
» just get_nodes
NAME           STATUS   ROLES                  AGE   VERSION
k3s-master     Ready    control-plane,master   86s   v1.26.1+k3s1
k3s-worker-1   Ready    worker                 32s   v1.26.1+k3s1

» just get_nodes
NAME           STATUS     ROLES                  AGE     VERSION
k3s-master     Ready      control-plane,master   3m52s   v1.26.1+k3s1
k3s-worker-1   NotReady   worker                 2m58s   v1.26.1+k3s1
```

I'll improve the code and try to make it work, I suspect the token is incorrectly propagated to the worker.

Up: Seems I ran the `just add_worker` too soon after the master. Waiting a couple of minutes first and then running it produced a worker which is `Ready` and stays that way.

**Comment by mihaigalos** [comment]
I dug deeper and think I got most, if not all of the files used by k3s in a tmpfs (see sources below).
This includes `/var/lib` and some volumes (such as the server folder) which were discussed in this thread.

Seems to be using about 215MB for the Master only, started off with an empty `/dev/shm`:
```
tmpfs              1.9G   214.8M     1.6G    11% ■■■□□□□□□□□□□□□□□□□□ /dev/shm
```

<details>
  <summary>Sources - click to expand!</summary>

```just
set shell := ["bash", "-uc"]

default:
    @just --list --unsorted

network := "k3s-home"
base_storage := "/dev/shm"
logs := base_storage + "/log/k3s"
lib := base_storage + "/lib/k3s"

# This is necessary do be done on host, since docker has no https for wget and no curl.
@_download_deps:
    [ ! -f argocd ] && wget https://github.com/argoproj/argo-cd/releases/latest/download/argocd-linux-arm64 -O argocd && chmod +x argocd || true

@_setup: && _download_deps
    [ -z $(sudo docker network ls --filter name=^{{ network }}$ --format="{{{{ .Name }}" ) ] && sudo docker network create k3s-home || true
    mkdir -p {{ lib }}
    mkdir -p {{ logs }}

add_master: _setup
    #!/bin/bash
    MASTER_NAME=k3s-master
    sudo docker run --rm -t \
        --add-host=host.docker.internal:host-gateway `# enable connect from external network`\
        --hostname $MASTER_NAME \
        --log-driver=none \
        --name=$MASTER_NAME \
        --net={{ network }} \
        --privileged `# TODO: remove` \
        -v $(pwd):/src \
        -v {{ lib }}/$MASTER_NAME/cni:/var/lib/cni \
        -v {{ lib }}/$MASTER_NAME/kubelet:/var/lib/kubelet \
        -v {{ lib }}/$MASTER_NAME/rancher_k3s:/var/lib/rancher/k3s \
        -v {{ logs }}/$MASTER_NAME:/var/log \
        rancher/k3s:latest server \
            --disable traefik \
            > {{ logs }}/$MASTER_NAME.log 2>&1 &
    disown

add_worker: _setup
    #!/bin/bash
    K3S_TOKEN=$(sudo docker exec -it k3s-master cat /var/lib/rancher/k3s/server/node-token | tr -d '\n' | tr -d '\r')
    WORKER_INDEX=$(( $(sudo docker ps | grep k3s-worker | wc -l) +1 ))
    WORKER_NAME=k3s-worker-$WORKER_INDEX
    sudo docker run --rm -t \
        --add-host=host.docker.internal:host-gateway `# enable connect from external network`\
        --log-driver=none \
        --name=$WORKER_NAME \
        --net={{ network }} \
        --privileged `# TODO: remove` \
        --tmpfs /run \
        --tmpfs /var/run \
        -e K3S_NODE_NAME=$WORKER_NAME \
        -e K3S_TOKEN=$K3S_TOKEN \
        -e K3S_URL=https://k3s-master:6443 \
        -v {{ lib }}/$WORKER_NAME/cni:/var/lib/cni \
        -v {{ lib }}/$WORKER_NAME/kubelet:/var/lib/kubelet \
        -v {{ lib }}/$WORKER_NAME/rancher_k3s:/var/lib/rancher/k3s \
        -v {{ logs }}/$WORKER_NAME:/var/log \
        rancher/k3s:latest agent \
            > {{ logs }}/$WORKER_NAME.log 2>&1 &
    disown
    sleep 3 # wait for registration of node in order to be able to wait for conditions pertaining to it.
    sudo docker exec -it k3s-master kubectl wait --for=condition=ready --timeout=300s node --all
    sudo docker exec -it k3s-master kubectl label node $WORKER_NAME kubernetes.io/role=worker

@get_nodes:
    sudo docker exec -it k3s-master kubectl get nodes

@kubectl +args="":
    sudo docker exec -it k3s-master kubectl {{ args }}
```
</details>

Up: I've noticed one can easily inspect I/O with htop (just press tab to switch between CPU and I/O):
![image](https://user-images.githubusercontent.com/16443090/236607436-41855272-d5a0-40a6-bbff-2cc437f28d3d.png)
